source/common/router/config_impl.h

```diff
@@ -335,6 +336,7 @@ class RouteEntryImplBase : public RouteEntry,
   }
   bool includeVirtualHostRateLimits() const override { return include_vh_rate_limits_; }
   const envoy::api::v2::core::Metadata& metadata() const override { return metadata_; }
+  const PathMatchCriterion& pathMatchCriterion() const override { return *this; }
```

Comment:
I'm wondering if I've forgotten how to read C++ :) Can you explain why this is needed (or even how it passes type checking?).

Reply:
Yeah, so the inheritance structure is a little convoluted here.  `RouteEntryImplBase` inherits from `RouteEntry`, `Matchable`, and now `PathMatchCriterion` (among others).  It has three derived classes (`PrefixRouteEntryImpl`, `PathRouteEntryImpl`, and `RegexRouteEntryImpl`) that implement the `Matchable` and `PathMatchCriterion` abstract methods since they implement all of the match-type specific logic.  So, for the `pathMatchCriterion()` method (part of the `RouteEntry` interface), it just returns itself as the `PathMatchCriterion` since it implements that interface.  Does that help clarify things?

Reply:
Does it ever need to be directly instantiated (i.e. be concrete)? Or do we only use derived classes? If the latter, it seems we don't need to define it.

Reply:
The intent of the class was to group the idea of a route match type and a matcher since one is needed to interpret the other (like a virtualized POD struct)  - as opposed to polluting the RouteEntry interface (which is already starting to feel a little big) with a `routeMatchType()` method and a `routeMatchString()` method or something.  I'm happy to promote them to the RouteEntry interface if that seems like a net gain, though.